boot/seal.go

```diff
@@ -25,65 +25,269 @@ import (
 	}
 
-	// Get the expected kernel command line of the recovery system we're installing from
-	recoveryCmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	runModeBootChains, err := buildRunModeBootChains(rbl, bl, model, modeenv, cmdline)
 	if err != nil {
-		return fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+		return fmt.Errorf("cannot build run mode boot chain: %v", err)
 	}
 
-	kernelCmdlines := []string{
-		cmdline,
-		recoveryCmdline,
+	pbc := toPredictableBootChains(append(runModeBootChains, recoveryBootChain))
+
+	// XXX: store the predictable bootchains
+
+	roleToBlName := map[bootloader.Role]string{
+		bootloader.RoleRecovery: rbl.Name(),
+		bootloader.RoleRunMode:  bl.Name(),
 	}
 
-	sealKeyParams := secboot.SealKeyParams{
-		ModelParams: []*secboot.SealKeyModelParams{
-			{
-				Model:          model,
-				KernelCmdlines: kernelCmdlines,
-				EFILoadChains:  []*secboot.LoadChain{loadChain},
-			},
-		},
-		KeyFile:                 filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
-		TPMPolicyUpdateDataFile: filepath.Join(InstallHostFDEDataDir, "policy-update-data"),
-		TPMLockoutAuthFile:      filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+	// get parameters from bootchains and seal the key
+	params, err := sealKeyParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot build key sealing parameters: %v", err)
 	}
 
-	if err := secbootSealKey(key, &sealKeyParams); err != nil {
+	if err := secbootSealKey(key, params); err != nil {
 		return fmt.Errorf("cannot seal the encryption key: %v", err)
 	}
 
 	return nil
 }
+
+func buildRecoveryBootChain(rbl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv) (bc bootChain, err error) {
+	// get the command line
+	cmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	if err != nil {
+		return bc, fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+	}
+
+	// get kernel information from seed
+	perf := timings.New(nil)
+	_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, modeenv.RecoverySystem, []snap.Type{snap.TypeKernel}, perf)
+	if err != nil {
+		return bc, err
+	}
+	if len(snaps) != 1 {
+		return bc, fmt.Errorf("cannot obtain recovery kernel snap")
+	}
+	seedKernel := snaps[0]
+
+	var kernelRev string
+	if seedKernel.SideInfo.Revision.Store() {
+		kernelRev = seedKernel.SideInfo.Revision.String()
+	}
+
+	// get asset chains
+	assetChain, kbf, err := buildRecoveryAssetChain(rbl, modeenv)
+	if err != nil {
+		return bc, err
+	}
+
+	return bootChain{
+		BrandID:        model.BrandID(),
+		Model:          model.Model(),
+		Grade:          string(model.Grade()),
+		ModelSignKeyID: model.SignKeyID(),
+		AssetChain:     assetChain,
+		Kernel:         seedKernel.Path,
+		KernelRevision: kernelRev,
+		KernelCmdlines: []string{cmdline},
+		model:          model,
+		kernelBootFile: bootloader.NewBootFile(seedKernel.Path, kbf.Path, kbf.Role),
```

Comment:
This should use `bootloader.NewBootFile(<snap-name>, ..)`

Reply:
no, Kernel: should use that, we need the file here, that's what sealing will read